include/alp/algorithms/symm_tridiag_eigensolver.hpp

```diff
@@ -0,0 +1,542 @@
+			rc = rc ? rc : alp::set( egvecs_view, tmp_egvecs_orth_view );
+
+			return rc;
+		}
+
+
+		/**
+		 * Calculate eigendecomposition of symmetric tridiagonal matrix T
+		 *        \f$T = Qdiag(d)Q^T\f$ where
+		 *        \a T is real symmetric tridiagonal
+		 *        \a Q is orthogonal (columns are eigenvectors).
+		 *        \a d is vector containing eigenvalues.
+		 *
+		 * @tparam D        Data element type
+		 * @tparam Ring     Type of the semiring used in the computation
+		 * @tparam Minus    Type of minus operator used in the computation
+		 * @tparam Divide   Type of divide operator used in the computation
+		 * @param[out] Q    output orthogonal matrix contaning eigenvectors
+		 * @param[out] d    output vector containg eigenvalues
+		 * @param[in]  T    input symmetric tridiagonal matrix
+		 * @param[in]  ring A semiring for operations
+		 * @return RC       SUCCESS if the execution was correct
+		 *
+		 */
+		template<
+			typename D,
+			typename SymmOrHermTridiagonalType,
+			typename OrthogonalType,
+			typename SymmHermTrdiViewType,
+			typename OrthViewType,
+			typename SymmHermTrdiImfR,
+			typename SymmHermTrdiImfC,
+			typename OrthViewImfR,
+			typename OrthViewImfC,
+			typename VecViewType,
+			typename VecImfR,
+			typename VecImfC,
+			class Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >,
+			class Minus = operators::subtract< D >,
+			class Divide = operators::divide< D >
+		>
+		RC symm_tridiag_dac_eigensolver(
+			Matrix<
+				D,
+				SymmOrHermTridiagonalType,
+				Dense,
+				SymmHermTrdiViewType,
+				SymmHermTrdiImfR,
+				SymmHermTrdiImfC
+			> &T,
+			Matrix<
+				D,
+				OrthogonalType,
+				Dense,
+				OrthViewType,
+				OrthViewImfR,
+				OrthViewImfC
+			> &Q,
+			Vector<
+				D,
+				structures::General,
+				Dense,
+				VecViewType,
+				VecImfR,
+				VecImfC
+			> &d,
+			const Ring &ring = Ring(),
+			const Minus &minus = Minus(),
+			const Divide &divide = Divide()
+		) {
+			(void) ring;
+			(void) minus;
+			(void) divide;
+
+			const Scalar< D > zero( ring.template getZero< D >() );
+			const Scalar< D > one( ring.template getOne< D >() );
+
+			RC rc = SUCCESS;
+
+			const size_t n = nrows( T );
+			const size_t m = n / 2;
+
+			if( n == 1 ) {
+				//d=T[0];
+				rc = rc ? rc : alp::eWiseLambda(
+					[ &d ]( const size_t i, const size_t j, D &val ) {
+						(void) i;
+						(void) j;
+						alp::set( d, Scalar< D > ( val ) );
+					},
+					T
+				);
+				// Q=[[1]]; a 1x1 matrix
+				rc = rc ? rc : alp::set( Q, one );
+
+				return rc;
+			}
+
+
+			Vector< D, structures::General, Dense > v( n );
+			rc = rc ? rc : alp::set( v, zero );
+			rc = rc ? rc : alp::eWiseLambda(
+				[ &T, &m, &ring ]( const size_t i, D &val ) {
+					if( i == m - 1 ) {
+						val = ring.template getOne< D >();
+					}
+					if( i == m) {
+						val = internal::access( T, internal::getStorageIndex( T, m - 1, m ) );
```

Comment:
We should not use internals in algorithms. Would `get_view` help here?

Reply:
Yes, I can rewrite it with views.

Reply:
I cannot capture type D if I completely get rid of view types.